File: backend/src/location/location.service.ts
import { Injectable } from '@nestjs/common';
import { getDistance } from 'geolib';

export interface UserLocation {
  userId: string;
  name: string; // User's name
  role: 'needy' | 'provider';
  latitude: number;
  longitude: number;
  isAvailable: boolean;
}

@Injectable()
export class LocationService {
  // In-memory storage (replace with database in production)
  private userLocations: Map<string, UserLocation> = new Map();

  calculateDistance(
    lat1: number,
    lon1: number,
    lat2: number,
    lon2: number,
  ): number {
    return getDistance(
      { latitude: lat1, longitude: lon1 },
      { latitude: lat2, longitude: lon2 },
    ) / 1000; // Convert to kilometers
  }

  findNearestProviders(
    needyLat: number,
    needyLon: number,
    maxDistanceKm: number = 10,
    limit: number = 5,
    excludeNeedyId?: string, // Optional: exclude this needy user's own ID
  ): UserLocation[] {
    const allLocations = Array.from(this.userLocations.values());
    const activeProviders = allLocations.filter(l => 
      l.role === 'provider' && 
      l.isAvailable && 
      (!excludeNeedyId || l.userId !== excludeNeedyId) // Exclude needy's own ID
    );
    
    const providers: UserLocation[] = [];

    for (const location of allLocations) {
      // Only show active providers, and exclude the needy user's own ID
      if (location.role === 'provider' && location.isAvailable && (!excludeNeedyId || location.userId !== excludeNeedyId)) {
        const distance = this.calculateDistance(
          needyLat,
          needyLon,
          location.latitude,
          location.longitude,
        );

        // Don't log individual provider distances - too verbose

        if (distance <= maxDistanceKm) {
          providers.push({
            ...location,
            // Add distance for sorting
            ...({ distance } as any),
          });
        }
      }
    }

    // Only log summary, not individual providers

    // Sort by distance and return top results
    return providers
      .sort((a, b) => (a as any).distance - (b as any).distance)
      .slice(0, limit);
  }

  findNearestNeedy(
    providerLat: number,
    providerLon: number,
    maxDistanceKm: number = 10,
    limit: number = 5,
    excludeProviderId?: string, // Optional: exclude this provider's own ID
  ): UserLocation[] {
    const allLocations = Array.from(this.userLocations.values());
    
    const activeNeedy = allLocations.filter(l => 
      l.role === 'needy' && 
      l.isAvailable && 
      (!excludeProviderId || l.userId !== excludeProviderId) // Exclude provider's own ID
    );
    
    // Only log if no needers found (important for debugging)
    if (activeNeedy.length === 0) {
      console.log('⚠️ No active needers found in system');
    }
    
    const needyUsers: UserLocation[] = [];

    for (const location of allLocations) {
      // Only show active needers, and exclude the provider's own ID
      if (location.role === 'needy' && location.isAvailable && (!excludeProviderId || location.userId !== excludeProviderId)) {
        const distance = this.calculateDistance(
          providerLat,
          providerLon,
          location.latitude,
          location.longitude,
        );

        // Don't log individual needy distances - too verbose

        if (distance <= maxDistanceKm) {
          needyUsers.push({
            ...location,
            ...({ distance } as any),
          });
        }
        // Don't log individual distances or summaries - too verbose
      }
    }

    return needyUsers
      .sort((a, b) => (a as any).distance - (b as any).distance)
      .slice(0, limit);
  }

  updateLocation(userId: string, location: UserLocation): void {
    // IMPORTANT: userId parameter is the backend-assigned user ID from signup
    // This ID is never changed - it's the same ID assigned during registration
    // Only location coordinates, name, and availability are updated here
    
    const existingLocation = this.userLocations.get(userId);
    // Only log on first registration or significant changes (role change, availability change)
    if (!existingLocation) {
      console.log(`📍 New user registered: ${location.role} ${location.name} (${userId})`);
    } else if (existingLocation.role !== location.role || existingLocation.isAvailable !== location.isAvailable) {
      console.log(`📍 User ${location.name} (${userId}) changed: role=${existingLocation.role}→${location.role}, available=${existingLocation.isAvailable}→${location.isAvailable}`);
    }
    // Don't log routine location updates - too verbose
    
    // Update location data but preserve the userId (it's the key in the map)
    // The userId in the location object should match the key
    this.userLocations.set(userId, { ...location, userId });
  }

  getLocation(userId: string): UserLocation | undefined {
    return this.userLocations.get(userId);
  }
}

File: backend-dotnet/Services/LocationService.cs
using FuelMateBackend.Models;

namespace FuelMateBackend.Services;

public class LocationService
{
    private readonly Dictionary<string, UserLocation> _userLocations = new();

    public double CalculateDistance(decimal lat1, decimal lon1, decimal lat2, decimal lon2)
    {
        // Haversine formula
        const double R = 6371; // Earth's radius in kilometers
        var dLat = ToRadians((double)(lat2 - lat1));
        var dLon = ToRadians((double)(lon2 - lon1));

        var a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2) +
                Math.Cos(ToRadians((double)lat1)) * Math.Cos(ToRadians((double)lat2)) *
                Math.Sin(dLon / 2) * Math.Sin(dLon / 2);

        var c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
        return R * c; // Distance in kilometers
    }

    private static double ToRadians(double degrees)
    {
        return degrees * Math.PI / 180;
    }

    public List<UserLocation> FindNearestProviders(
        decimal needyLat, 
        decimal needyLon, 
        double maxDistanceKm = 10, 
        int limit = 5,
        string? excludeNeedyId = null)
    {
        var providers = _userLocations.Values
            .Where(l => l.Role == "provider" && 
                       l.IsAvailable && 
                       (excludeNeedyId == null || l.UserId != excludeNeedyId))
            .Select(location => {
                var distance = CalculateDistance(needyLat, needyLon, location.Latitude, location.Longitude);
                return new UserLocation
                {
                    UserId = location.UserId,
                    Name = location.Name,
                    Role = location.Role,
                    Latitude = location.Latitude,
                    Longitude = location.Longitude,
                    IsAvailable = location.IsAvailable,
                    Distance = distance
                };
            })
            .Where(p => p.Distance <= maxDistanceKm)
            .OrderBy(p => p.Distance)
            .Take(limit)
            .ToList();

        return providers;
    }

    public List<UserLocation> FindNearestNeedy(
        decimal providerLat, 
        decimal providerLon, 
        double maxDistanceKm = 10, 
        int limit = 5,
        string? excludeProviderId = null)
    {
        var needyUsers = _userLocations.Values
            .Where(l => l.Role == "needy" && 
                       l.IsAvailable && 
                       (excludeProviderId == null || l.UserId != excludeProviderId))
            .Select(location => {
                var distance = CalculateDistance(providerLat, providerLon, location.Latitude, location.Longitude);
                return new UserLocation
                {
                    UserId = location.UserId,
                    Name = location.Name,
                    Role = location.Role,
                    Latitude = location.Latitude,
                    Longitude = location.Longitude,
                    IsAvailable = location.IsAvailable,
                    Distance = distance
                };
            })
            .Where(n => n.Distance <= maxDistanceKm)
            .OrderBy(n => n.Distance)
            .Take(limit)
            .ToList();

        if (needyUsers.Count == 0)
        {
            Console.WriteLine("⚠️ No active needers found in system");
        }

        return needyUsers;
    }

    public void UpdateLocation(string userId, UserLocation location)
    {
        var existingLocation = GetLocation(userId);
        
        if (existingLocation == null)
        {
            Console.WriteLine($"📍 New user registered: {location.Role} {location.Name} ({userId})");
        }
        else if (existingLocation.Role != location.Role || existingLocation.IsAvailable != location.IsAvailable)
        {
            Console.WriteLine($"📍 User {location.Name} ({userId}) changed: role={existingLocation.Role}→{location.Role}, available={existingLocation.IsAvailable}→{location.IsAvailable}");
        }

        _userLocations[userId] = new UserLocation
        {
            UserId = userId,
            Name = location.Name,
            Role = location.Role,
            Latitude = location.Latitude,
            Longitude = location.Longitude,
            IsAvailable = location.IsAvailable
        };
    }

    public UserLocation? GetLocation(string userId)
    {
        return _userLocations.TryGetValue(userId, out var location) ? location : null;
    }
}

